### DISCUSSION TITLE
use addional arguments when writing raster / define own driver

### DISCUSSION THREAD

**Discussion by wiesehahn** [r-lidar/lidR#589] [Q&A]
with `terra::writeRaster()` it is possible to use additional arguments. 
Let's say I want to pass some of these arguments (e.g. `gdal=c("COMPRESS=DEFLATE")`) when applying a function on a catalog where the output should be stored on disk. I guess I cannot pass these arguments directly but have to initialize my own driver!?

I tried to to do this but the function seems to use the standard raster driver, is there a way to define which driver to use?

```r
write_rast = function(rast, file)
{
  file1 = tools::file_path_sans_ext(file)
  file1 = paste0(file1, "_test.tif")
  terra::writeRaster(rast, file1, gdal=c("COMPRESS=DEFLATE"))
}



some_function <- function(las, resolution = 1)
{
  if (is(las, "LAScatalog"))  {
    las@output_options$drivers$list = list(
      write = write_rast,
      object = "rast",
      path = "file",
      extension = "",
      param = list())
    options <- list(automerge = FALSE, need_buffer = TRUE)
    return(catalog_apply(las, some_function, resolution=resolution, .options = options))
  }
  else if (is(las, "LAScluster")) {
    las <- readLAS(las)
    if (is.empty(las)) return(NULL) 
    dat <- some_function(las, resolution=resolution)
    dat <- terra::crop(dat, ext(chunk))
    return(dat)
  }
  else if (is(las, "LAS")) {
    dat <- rasterize_terrain(las, resolution, tin(), pkg = "terra")
    
    return(dat)
  }
  else {
    stop("This type is not supported.")
  }
}

some_function(ctg, 5)

```

**Comment by Jean-Romain** [comment]
Are you looking for

```r
ctg@output_options$drivers$SpatRaster$param$gdal = c("COMPRESS=DEFLATE")
```

**Reply by wiesehahn** [reply]
Yes and no, as I described it this is exactly what I was trying to do. 
But there are also options which are not supported by terra. For example it does not support WEBP-compression. The solution I was thinking about is not very nice but it would extend the write_rast() function to first write the raster and then call a function to run gdal_translate.

How would I tell my function to use this specific driver instead of the standard?

**Reply by Jean-Romain** [reply]
```r
las@output_options$drivers$SpatRaster = list(
      write = write_rast,
      object = "rast",
      path = "file",
      extension = "",
      param = list())
```
